### PR TITLE
Add webhook for ContainerRecreateRequest

### PR DESCRIPTION
**Pull request by FillZpp**
Signed-off-by: FillZpp <FillZpp.pub@gmail.com>

### Ⅰ. Describe what this PR does
Add webhook for ContainerRecreateRequest

Reference to proposal [ContainerRecreateRequest](https://github.com/openkruise/kruise/blob/master/docs/proposals/20210316-containerrecreaterequest.md).

### CODE REVIEW

File: pkg/webhook/containerrecreaterequest/mutating/crr_mutating_handler.go [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	if req.AdmissionRequest.Operation == admissionv1beta1.Update {
+		oldObj := &appsv1alpha1.ContainerRecreateRequest{}
+		if err := h.Decoder.DecodeRaw(req.OldObject, oldObj); err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+		if !reflect.DeepEqual(obj.Spec, oldObj.Spec) {
+			return admission.Errored(http.StatusForbidden, fmt.Errorf("not allowed to update spec of ContainerRecreateRequest"))
```
Comment (furykerry) [inline pkg/webhook/containerrecreaterequest/mutating/crr_mutating_handler.go:69]: "spec of ContainerRecreateRequest is immutable"

File: pkg/webhook/containerrecreaterequest/mutating/crr_mutating_handler.go [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+			return fmt.Errorf("can not recreate %s multi times", c.Name)
+		}
+
+		if c.PreStop != nil || c.Ports != nil || c.StatusContext != nil {
+			return fmt.Errorf("preStop, ports, statusContext in container are ready-only fields")
+		}
+
+		podContainer := util.GetContainer(c.Name, pod)
+		if podContainer == nil {
+			return fmt.Errorf("not found %s container in Pod", c.Name)
```
Comment (furykerry) [inline pkg/webhook/containerrecreaterequest/mutating/crr_mutating_handler.go:163]: container %s not found in Pod

**Review by furykerry — CHANGES_REQUESTED** [review]

**Review by Fei-Guo — COMMENTED** [review]
LGTM

File: pkg/webhook/containerrecreaterequest/mutating/crr_mutating_handler.go [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+		if err := h.Decoder.DecodeRaw(req.OldObject, oldObj); err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+		if !reflect.DeepEqual(obj.Spec, oldObj.Spec) {
+			return admission.Errored(http.StatusForbidden, fmt.Errorf("spec of ContainerRecreateRequest is immutable"))
+		}
+		if obj.Labels[appsv1alpha1.ContainerRecreateRequestPodNameKey] != oldObj.Labels[appsv1alpha1.ContainerRecreateRequestPodNameKey] ||
+			obj.Labels[appsv1alpha1.ContainerRecreateRequestPodUIDKey] != oldObj.Labels[appsv1alpha1.ContainerRecreateRequestPodUIDKey] ||
+			obj.Labels[appsv1alpha1.ContainerRecreateRequestNodeNameKey] != oldObj.Labels[appsv1alpha1.ContainerRecreateRequestNodeNameKey] {
+			return admission.Errored(http.StatusForbidden, fmt.Errorf("not allowed to update immutable labels of ContainerRecreateRequest"))
```
Comment (furykerry) [inline pkg/webhook/containerrecreaterequest/mutating/crr_mutating_handler.go:74]: give the specific immutable label in the error for better user understanding

Comment (FillZpp) [reply pkg/webhook/containerrecreaterequest/mutating/crr_mutating_handler.go]: done

**Review by furykerry — APPROVED** [review]

**Pull request merged**